File: tests/e2e/nodejs_client/test/admin.test.js
/**
 * E2E: Phase 10 admin endpoints against a live bqemulator container
 * via raw HTTP requests (the Node BigQuery client doesn't expose the
 * /admin/* surface, so we use fetch).
 *
 * Ship criterion: ``bqemulator import/export/seed/backup/restore`` all
 * round-trip. The four non-admin commands are CLI-only and exercised
 * by the Python E2E + integration suites. The Node E2E covers the
 * /admin/* JSON surface against the same live container.
 */

const { describe, it, before, after } = require("node:test");
const assert = require("node:assert/strict");

const REST_URL = process.env.BQEMU_REST_URL || "http://localhost:9050";
const PROJECT = "e2e-nodejs-admin";
const DATASET = "admin_node_ds";

async function rest(method, path, body) {
  const res = await fetch(REST_URL + path, {
    method,
    headers: { "Content-Type": "application/json" },
    body: body ? JSON.stringify(body) : undefined,
  });
  if (!res.ok && method !== "DELETE") {
    throw new Error(`${method} ${path} → ${res.status}: ${await res.text()}`);
  }
  return res.status === 204 ? null : res.json().catch(() => null);
}

describe("Phase 10 — admin endpoints (live container)", () => {
  before(async () => {
    await rest("POST", `/bigquery/v2/projects/${PROJECT}/datasets`, {
      datasetReference: { projectId: PROJECT, datasetId: DATASET },
    });
  });

  after(async () => {
    await rest(
      "DELETE",
      `/bigquery/v2/projects/${PROJECT}/datasets/${DATASET}?deleteContents=true`,
    );
  });

  it("GET /admin/config returns settings with admin_enabled=true", async () => {
    const body = await rest("GET", "/admin/config");
    if (!body) return; // admin may be disabled — Python E2E will skip
    assert.equal(body.kind, "bqemu#adminConfig");
    assert.equal(body.settings.admin_enabled, true);
  });

  it("GET /admin/catalog reports the seeded dataset", async () => {
    await rest("POST", `/bigquery/v2/projects/${PROJECT}/datasets/${DATASET}/tables`, {
      schema: { fields: [{ name: "n", type: "INT64", mode: "REQUIRED" }] },
      tableReference: {
        projectId: PROJECT,
        datasetId: DATASET,
        tableId: "t1",
      },
    });
    const body = await rest("GET", `/admin/catalog?projectId=${PROJECT}`);
    if (!body) return;
    assert.equal(body.kind, "bqemu#adminCatalog");
    const proj = body.projects.find((p) => p.projectId === PROJECT);
    assert.ok(proj, "project should appear in admin catalog");
    const ds = proj.datasets.find((d) => d.datasetId === DATASET);
    assert.ok(ds, "dataset should appear");
    assert.ok(
      ds.tables.some((t) => t.tableId === "t1"),
      "table t1 should appear",
    );
  });

  it("GET /admin/jobs returns a job list after a query", async () => {
    await rest("POST", `/bigquery/v2/projects/${PROJECT}/jobs`, {
      configuration: {
        query: { query: "SELECT 1", useLegacySql: false },
      },
    });
    const body = await rest("GET", `/admin/jobs?projectId=${PROJECT}`);
    if (!body) return;
    assert.equal(body.kind, "bqemu#adminJobList");
    assert.ok(body.totalItems >= 1);
  });

  it("GET /admin/streams returns counts", async () => {
    const body = await rest("GET", "/admin/streams");
    if (!body) return;
    assert.equal(body.kind, "bqemu#adminStreamList");
    assert.ok("writeStreamCount" in body);
    assert.ok("readSessionCount" in body);
  });
});

File: tests/e2e/go_client/admin_test.go
// Phase 10 admin endpoints E2E via raw HTTP against a live bqemulator
// container.
//
// Ship criterion: ``bqemulator import/export/seed/backup/restore`` all
// round-trip cleanly; CLI-only paths are exercised by the Python E2E +
// integration suites. This file covers the /admin/* JSON surface against
// the same live container the rest of the Go E2E suite hits.

package e2e

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"testing"
)

const adminProject = "e2e-go-admin"

func adminRest(t *testing.T, method, path string, body any) (int, map[string]any) {
	t.Helper()
	var buf io.Reader
	if body != nil {
		raw, _ := json.Marshal(body)
		buf = bytes.NewReader(raw)
	}
	req, err := http.NewRequest(method, restURL()+path, buf)
	if err != nil {
		t.Fatalf("rest %s %s: %v", method, path, err)
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatalf("rest %s %s: %v", method, path, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode == 204 {
		return resp.StatusCode, nil
	}
	if resp.StatusCode >= 400 && method != "DELETE" && path != "/admin/config" {
		body, _ := io.ReadAll(resp.Body)
		t.Fatalf("rest %s %s: %d %s", method, path, resp.StatusCode, string(body))
	}
	var out map[string]any
	_ = json.NewDecoder(resp.Body).Decode(&out)
	return resp.StatusCode, out
}

func TestAdminAdminEndpoints(t *testing.T) {
	dataset := "admin_go_ds"
	adminRest(t, "POST",
		fmt.Sprintf("/bigquery/v2/projects/%s/datasets", adminProject),
		map[string]any{
			"datasetReference": map[string]any{
				"projectId": adminProject,
				"datasetId": dataset,
			},
		})
	defer adminRest(t, "DELETE",
		fmt.Sprintf("/bigquery/v2/projects/%s/datasets/%s?deleteContents=true",
			adminProject, dataset), nil)

	// /admin/config — admin must be enabled for this to be 200.
	status, body := adminRest(t, "GET", "/admin/config", nil)
	if status == 404 {
		t.Skip("admin disabled in container; covered by unit tests")
	}
	if got, want := body["kind"], "bqemu#adminConfig"; got != want {
		t.Errorf("/admin/config kind: got %v want %v", got, want)
	}

	// /admin/catalog reports the dataset.
	_, body = adminRest(t, "GET", "/admin/catalog?projectId="+adminProject, nil)
	if body["kind"] != "bqemu#adminCatalog" {
		t.Errorf("/admin/catalog kind: %v", body["kind"])
	}

	// /admin/jobs returns a list.
	_, body = adminRest(t, "GET", "/admin/jobs", nil)
	if body["kind"] != "bqemu#adminJobList" {
		t.Errorf("/admin/jobs kind: %v", body["kind"])
	}

	// /admin/streams returns stream counts.
	_, body = adminRest(t, "GET", "/admin/streams", nil)
	if body["kind"] != "bqemu#adminStreamList" {
		t.Errorf("/admin/streams kind: %v", body["kind"])
	}
}
